File: frontend/src/api/alerts.ts
import request from './request';

export interface AlertRuleType {
  change_percent?: { threshold: number };
  sealed_amount?: { threshold: number };
  intensity_score?: { threshold: number };
  anomaly?: { volume_change_ratio: number; price_change_ratio: number };
}

export interface AlertRule {
  id: string;
  name: string;
  rule_type: 'change_percent' | 'sealed_amount' | 'intensity_score' | 'anomaly';
  rule_config: AlertRuleType;
  enabled: boolean;
  created_at: string;
}

export interface AlertEvent {
  id: string;
  rule_id: string;
  rule_name: string;
  stock_code: string;
  stock_name: string;
  message: string;
  severity: 'info' | 'warning' | 'critical';
  triggered_at: string;
  metadata: {
    price: number;
    change_percent: number;
    sealed_amount_buy: number;
    sealed_amount_sell: number;
  };
}

// 获取告警规则列表
export async function getAlertRules(): Promise<AlertRule[]> {
  const response = await request.get<{rules: AlertRule[]}>('/auction/alerts');
  return response.rules || [];
}

// 创建告警规则
export async function createAlertRule(
  name: string,
  ruleType: AlertRuleType,
  enabled: boolean = true
): Promise<AlertRule> {
  return request.post('/auction/alerts', {
    name,
    rule_type: ruleType,
    enabled,
  });
}

// 删除告警规则
export async function deleteAlertRule(ruleId: string): Promise<void> {
  return request.delete(`/auction/alerts/${ruleId}`);
}

// 获取告警历史
export async function getAlertHistory(limit: number = 100): Promise<AlertEvent[]> {
  const response = await request.get<{alerts: AlertEvent[]}>(
    `/auction/alerts/history?limit=${limit}`
  );
  return response.alerts || [];
}

File: services/auction-storage/src/domain/entities/alerts.rs
//! 告警相关实体

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// 告警规则类型
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AlertRuleType {
    /// 价格涨幅告警
    ChangePercent { threshold: f64 },
    /// 封单金额告警（万元）
    SealedAmount { threshold: f64 },
    /// 强度评分告警
    IntensityScore { threshold: f32 },
    /// 异动告警（成交量或价格突然变化）
    Anomaly {
        volume_change_ratio: f64,
        price_change_ratio: f64,
    },
}

/// 告警规则
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlertRule {
    pub id: String,
    pub name: String,
    pub rule_type: AlertRuleType,
    pub enabled: bool,
    pub created_at: DateTime<Utc>,
}

/// 告警事件
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AlertEvent {
    pub id: String,
    pub rule_id: String,
    pub rule_name: String,
    pub stock_code: String,
    pub stock_name: String,
    pub message: String,
    pub severity: AlertSeverity,
    pub triggered_at: DateTime<Utc>,
    pub metadata: serde_json::Value,
}

/// 告警严重程度
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum AlertSeverity {
    Info,
    Warning,
    Critical,
}
